:memo: Adding Steps for 09-recurssion tutorial (#796)
o1-labs/docs2: docs/zkapps/tutorials/09-recursion.mdx

Proofs generated using a ZkProgram can be passed into zkApp smart contracts for them to verify recursively. They can even be passed recursively into their own functions for off-chain recursive composition.

> The full source code for the following ZkProgram tutorial is provided in the [main.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/main.ts) directory on GitHub. While you're there, give the `/docs2` repository a star so that other zk developers can learn to build a zkApp!

Reviewer: ```suggestion
The following example code for the ZkProgram tutorial is provided in the [main.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/main.ts) file on GitHub. While you're there, give the `/docs2` repository a star so that other zk developers can learn to build a zkApp!
```

## Prerequisites

Ensure your environment meets the [Prerequisites](/zkapps/tutorials#prerequisites) for zkApp Developer Tutorials.

In particular, make sure you have the zkApp CLI installed:

```sh
$ npm install -g zkapp-cli
```

## Create a new project

Now that you have the tooling installed, you can start building your application.

1. Create or change to a directory where you have write privileges.
1. Now, create a project using the `zk project` command:

> 
> ```sh
> $ zk project 09-recursion
> ```
> 
> As you learned in earlier tutorials, the `zk project` command creates the `09-recursion` directory that contains the scaffolding for your project.

Reviewer: ```suggestion

    ```sh
    $ zk project 09-recursion
    ```
    
    As you learned in earlier tutorials, the `zk project` command creates the `09-recursion` directory that contains the scaffolding for your project.
```

Reviewer: @LuffySama-Dev to keep enumeration of the Markdown numbered list, we indent the content for each step, see https://github.com/o1-labs/docs2/wiki/Markdown-Syntax#numbered-steps

1. Change into the `09-recursion` directory. Like all projects, you run commands from the root of the `09-recursion` directory as you work in the `src` directory on files that contain the TypeScript code for the smart contract. Each time you make updates, then build or deploy, the TypeScript code is compiled into JavaScript in the `build` directory.

### Prepare the project

> Like earlier tutorials, you can prepare your project by deleting the default files that come with the new project and create new file called Add.

Reviewer: ```suggestion
Like earlier tutorials, you can prepare your project by deleting the default files that come with the new project and creating a smart contract called `Add`.
```

> Open `src/index.ts` in a text editor and import the Add smart contract, like:

Reviewer: ```suggestion
Open `src/index.ts` in a text editor and import the `Add` smart contract, like:
```

	
	```ts src/index.ts
	import { Add } from './Add.js';
	
	export { Add };

## Write the ZkProgram

Now, the fun part! Write your smart contract in the `src/Add.ts` file.

A final version of the smart contract is provided in the [Add.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/Add.ts) example file.

### Copy the example

> Use the Add code already completed in the [Add.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/Add.ts) example file.

Reviewer: ```suggestion
Use the existing code in the [Add.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/Add.ts) example file.
```

1. First, open the [Add.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/Add.ts) example file.
	
1. Copy the file's entire contents into your project `src/Add.ts` file.

### Imports

The `import` statement brings in other packages and dependencies to use in your ZkProgram.

:::info

All functions used inside a ZkProgram must operate on o1js compatible data types: `Field` types and other types built on top of `Field` types.

:::info

```ts src/Add.ts
import { Field, SelfProof, ZkProgram, verify } from 'o1js';
```

These items are:

> - `Field`: The native number type in o1js. You can think of Field elements as unsigned integers. Field elements are the most basic type in o1js. All other o1js-compatible types are built on top of Field elements.
> - `SelfProof`: The class in o1js which extends Proof class. It allows user to pass in a Proof of a Circuit in one of it's own methods.
> - `ZkProgram`: The o1js general purpose API for creating zero knowledge proofs. A ZkProgram is similar to zkApp smart contracts but isn't tied to an on-chain account.
> - `verify`: Verifies the Signature using a message and the corresponding PublicKey.

Reviewer: ```suggestion
- `Field`: The native number type in o1js. You can think of `Field` elements as unsigned integers. Field elements are the most basic type in o1js. All other o1js-compatible types are built on top of `Field` elements.
- `SelfProof`: The class in o1js that extends the `Proof` class and allows you to pass in a proof of a circuit in one of its own methods.
- `ZkProgram`: The o1js general purpose API for creating zero knowledge proofs. A ZkProgram is similar to zkApp smart contracts but isn't tied to an on-chain account.
- `verify`: Verifies the signature using a message and the corresponding `PublicKey`.
```

Recursive zero knowledge proofs can help you build powerful zkApps.

> Check out [Tutorial 10: Account Updates](/zkapps/tutorials/account-updates) to learn about account updates, the underlying structure of zkApps, and how they enable permissions, preconditions, and composability.

Reviewer: @LuffySama-Dev  for some reason, this link is breaking Vercel CI

Do you have the entire docs2 repo cloned for this fork? 

Tip: See https://github.com/o1-labs/docs2/blob/main/README.md#how-to-see-your-local-doc-updates-during-development 

and https://github.com/o1-labs/docs2/blob/main/README.md#debugging-build-errors 

This link is valid, but the PR cannot move forward with failing CI.

Author: Okay, I will check and update here before EOD.

Author: Hi @barriebyron ,

I tried to check this locally and link is fine, but the build is failing.
I am checking why is it happening as only line number changed an rest everything is same.

Thank you

Author: Hi @barriebyron ,

It seems in one of the recent PR #839 , a extra `(` was added in `on-chain-values.mdx`.
It was creating issues for us during build.
I have removed the extra parenthesis and build is successful locally.
I am committing it right away.

![Screenshot 2024-02-13 at 11 38 54 PM](https://github.com/o1-labs/docs2/assets/55076843/dd3265a6-1772-4633-b8a8-041a539e8b1c)

Thank You 😃

Reviewer: yes you're right, per usual! thank you, fixed on main
